app/src/main/java/com/ecchilon/sadpanda/api/DataLoader.java

```diff
@@ -51,252 +52,288 @@
+			Request.Builder builder = new Request.Builder()
+					.addHeader("Cookie", auth.getSessionCookie())
+					.url(url)
+					.get();
+			if (!useCache) {
+				builder.cacheControl(CacheControl.FORCE_NETWORK);
 			}
-
-			Matcher matcher = pShowkey.matcher(content);
-			String showkey = "";
-
-			while (matcher.find()) {
-				showkey = matcher.group(1);
+			String content;
+			try {
+				Response response = client.newCall(builder.build()).execute();
+				content = response.body().string();
 			}
-
-			if (Strings.isEmpty(showkey)) {
-				throw new ApiCallException(ApiErrorCode.SHOWKEY_NOT_FOUND, url, response);
+			catch (IOException e) {
+				if (!subscriber.isUnsubscribed()) {
+					subscriber.onError(new ApiCallException(ApiErrorCode.IO_ERROR, e));
+				}
+				return;
 			}
-			else {
-				gallery.setShowkey(showkey);
 
-				return showkey;
+			if (!subscriber.isUnsubscribed()) {
+				subscriber.onNext(content);
+				subscriber.onCompleted();
 			}
-		}
-		catch (IOException e) {
-			throw new ApiCallException(ApiErrorCode.IO_ERROR, e);
-		}
+		}).subscribeOn(Schedulers.io());
 	}
 
-	public List<GalleryEntry> getGalleryIndex(String base, boolean cache) throws ApiCallException {
+	public Observable<List<GalleryEntry>> getGalleryIndex(String base, boolean cache) {
 		return getGalleryIndex(base, 0);
 	}
 
-	public List<GalleryEntry> getGalleryIndex(String base, int page) throws ApiCallException {
+	public Observable<List<GalleryEntry>> getGalleryIndex(String base, int page) {
 		return getGalleryIndex(base, page, true);
 	}
 
-	public List<GalleryEntry> getGalleryIndex(String base, int page, boolean cache) throws ApiCallException {
-		String url = getGalleryIndexUrl(base, page);
-
-		try {
-			HttpGet httpGet = new HttpGet(url);
-			if (!cache) {
-				httpGet.addHeader("Cache-Control", "no-cache");
-			}
-			HttpResponse response = getHttpResponse(httpGet);
-			String html = readResponse(response);
-			Matcher matcher = pGalleryHref.matcher(html);
-			JSONArray gidlist = new JSONArray();
+	public Observable<List<GalleryEntry>> getGalleryIndex(String base, int page, boolean cache) {
+		return getContent(getGalleryIndexUrl(base, page), cache)
+				.flatMap(content -> {
+					Matcher matcher = pGalleryHref.matcher(content);
+					JSONArray gidlist = new JSONArray();
 
-			while (matcher.find()) {
-				long id = Long.parseLong(matcher.group(2));
-				String token = matcher.group(3);
-				JSONArray arr = new JSONArray();
+					while (matcher.find()) {
+						long id = Long.parseLong(matcher.group(2));
+						String token = matcher.group(3);
+						JSONArray arr = new JSONArray();
 
-				arr.put(id);
-				arr.put(token);
+						arr.put(id);
+						arr.put(token);
 
-				gidlist.put(arr);
-			}
+						gidlist.put(arr);
+					}
 
-			return getGalleryList(gidlist);
-		}
-		catch (IOException e) {
-			throw new ApiCallException(ApiErrorCode.IO_ERROR, e);
-		}
+					return getGalleryList(gidlist);
+				});
```

Comment:
`.subscribeOn(Schedulers.computation())`? Might be faster since the gallery matching might take a bit.